web/ui/react-app/src/Graph.tsx

```diff
@@ -3,94 +3,86 @@ import React, { PureComponent } from 'react';
 import ReactResizeDetector from 'react-resize-detector';
 import { Alert } from 'reactstrap';
 
-import Legend from './Legend';
 import { escapeHTML } from './utils/html';
-
+import SeriesName from './SeriesName';
+import { Nullable, GraphSeries, GraphMetric, QueryParams } from './types/types';
 require('flot');
 require('flot/source/jquery.flot.crosshair');
 require('flot/source/jquery.flot.legend');
 require('flot/source/jquery.flot.time');
 require('flot/source/jquery.canvaswrapper');
 require('jquery.flot.tooltip');
 
-let graphID = 0;
-function getGraphID() {
-  // TODO: This is ugly.
-  return graphID++;
-}
-
 interface GraphProps {
-  data: any; // TODO: Type this.
+  data: {
+    resultType: string;
+    result: Array<{ metric: GraphMetric; values: [number, string][] }>;
+  };
   stacked: boolean;
-  queryParams: {
-    startTime: number;
-    endTime: number;
-    resolution: number;
-  } | null;
+  queryParams: Nullable<QueryParams>;
 }
 
-class Graph extends PureComponent<GraphProps> {
-  private id: number = getGraphID();
+interface GraphState {
+  selectedSeriesIndex: Nullable<number>;
+  hoveredSeriesIndex: Nullable<number>;
+}
+
+class Graph extends PureComponent<GraphProps, GraphState> {
   private chartRef = React.createRef<HTMLDivElement>();
 
-  renderLabels(labels: { [key: string]: string }) {
-    const labelStrings: string[] = [];
-    for (const label in labels) {
-      if (label !== '__name__') {
-        labelStrings.push('<strong>' + label + '</strong>: ' + escapeHTML(labels[label]));
-      }
-    }
-    return '<div class="labels">' + labelStrings.join('<br>') + '</div>';
-  }
+  state = {
+    selectedSeriesIndex: null,
+    hoveredSeriesIndex: null,
+  };
 
-  formatValue = (y: number | null): string => {
+  formatValue = (y: Nullable<number>): string => {
     if (y === null) {
       return 'null';
     }
-    const abs_y = Math.abs(y);
-    if (abs_y >= 1e24) {
+    const absY = Math.abs(y);
```

Comment:
Heh yeah, these somehow made it over from the old code with old naming :P Thanks for fixing.

Reply:
Yeah, and eslint it would take my eyes out :D